src/lib/registration/widgets/checkbox_list.rb

```diff
@@ -0,0 +1,70 @@
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "cwm/common_widgets"
+
+module Registration
+  module Widgets
+    # A check boxes list intended to be the master widget in {MasterDetailSelector}
+    class CheckboxList < CWM::RichText
```

Comment:
I do not like this class at all :) It pretends to be checkbox list, but it is in fact just richtext box which remember scrolling position.

Reply:
Yes, this a convenience and quite badly named widget. It's here to avoid adding the logic to update (`Yast::UI.ChangeWidget(Id(...), :Value, new_content)`) and remember the scrolling of the _mater area_ directly from the `MasterDetailSelector` widget.

But maybe we should simply move it there, although I must admit that I don't like too much the idea. 

Could you please help me to find a better solution? (taking into account that we are looking for a reusable multi-state selector widget).

I will be extremely grateful to you for doing so.

Reply:
OK, so what responsibility/features we need here?

1. remember scrolling area - This should in my opinion should go to CWM::RichText itself as flag and feature as other richtext area can benefit from it. ( so basically if flag is set modify value= method ).

2. checkboxes representation in Richtext - it is now in Checkbox Item. It looks fine for me. But what I do not like is that this widget is stupid and get just content. I think it should get list of Checkbox Item and render it itself like e.g. Tabs or Trees do it.

3. reaction to click so it changes states ( like default action that button will changed to selected/unselected ). I think this definitively should be handled here. As part of it I see also method that recognize if event belongs to widget.

4. updating whole widget as there is auto dependencies between addons. This for me belongs to non-generic AddonsList which is child of CheckBox list and which has own handle that do this redrawing and recomputing. This is usual CWM way. You are building bricks of widgets together and us the widget that fits you best.

Reply:
First of all, thanks for the detailed answer @jreidinger, but I fear I do not completely agree with that vision

Probably I should have started commenting that those widgets are stupid on purpose. They are there only to "represent" the information, not to deal with it. They have no other responsibility. They know nothing about the selection logic or selected items. They have no state.

That's the way I found more appropriate to make them reusable. And that's why the third player, the controller, is needed. It is its responsibility to provide the information and deal with it.

Then, in this PR we have

* A main/wrapper/container widget (aka `MasterDetailSelector`): it has no state either, it's the only responsibility is react to certain events, requesting needed actions to the controller and keeping its child widgets updated and in sync.
* The list container (aka `CheckboxList`): will contain all checkboxes items and provides the single #update method.
* The detail container (aka `ItemDetails`): mostly the same as the list, but for the details of the currently selected item.
* The controller: is the key piece here. It knows where to look for the information, how to make the selection and auto-selection, etc. It should provide all necessary information and actions needed by the main widget.

Of course, those two child widgets are only convenience widgets, we can do the same directly from the `MasterDetailSelector` but I thought it would be a good idea to keep the main widget agnostic enough: in the end, it only needs to call  `#update` or `#reset` method. The rest of the story (saving/restoring the scroll, the place holder, etc) is not its business.

I don't know if I explained it better now or if that philosophy does not fit with CWM, but that's the approach I wanted to follow: making a _representational widgets_ that only knows how to present the information to the user and how to ask for it. Nothing more.

Do you still thinking the widgets need to be stateful? If so, are you convinced they still being reusable for other use cases?

Apart from those questions, do you think we could at least give it a try in `yast2-registration` before refining them and move it to a more centric place?

Thanks.

Reply:
> This should in my opinion should go to CWM::RichText itself as flag and feature as other richtext area can benefit from it

That's a good idea. So, if I got it correctly, we should fix it [over here](https://github.com/yast/yast-yast2/blob/b0eebf01c5485448393933563082cb908de0d3fc/library/cwm/src/lib/cwm/common_widgets.rb#L364-L368) make the PR and then update the [yast2 dependency](https://github.com/yast/yast-registration/blob/9c04bc3d0c44488425f056f781d0c091a9efb10e/package/yast2-registration.spec#L43) here, right?

Reply:
> > This should in my opinion should go to CWM::RichText itself as flag and feature as other richtext area can benefit from it
> 
> That's a good idea. So, if I got it correctly, we should fix it in cwm/common_widgets ([over here](https://github.com/yast/yast-yast2/blob/b0eebf01c5485448393933563082cb908de0d3fc/library/cwm/src/lib/cwm/common_widgets.rb#L364-L368)), make the PR and then update the [yast2 dependency](https://github.com/yast/yast-registration/blob/9c04bc3d0c44488425f056f781d0c091a9efb10e/package/yast2-registration.spec#L43) here, right?

yes

Reply:
@dgdavid regarding that widget, I find it very inconsistent with rest of other widgets regarding selecting/unselecting checkboxes. For other widgets and I mean mainly https://github.com/yast/yast-yast2/blob/b0eebf01c5485448393933563082cb908de0d3fc/library/cwm/src/lib/cwm/common_widgets.rb#L214 which is the closest to this checkbox list you can easily set list of selected items and also easily get info what is selected in that widget. See also similar table widget. So what I do not like is inconsistency with other widgets.

Also as I said consider you want to reuse your code aka widgets. What it means? you can use widgets which does not provide much value. You can reuse list container and can reuse details container.

I think that details widget can be used standalone without issues.
But can you reuse standlone CheckboxList? I think it does not provide much value. So if I want to reuse it I am force to use MasterDetailSelector, but what if I do not want details? Then I have to write it again from scratch and CheckboxList itself does not provide much.

Also consider what provides standalone CheckboxItem. Does it make sense to use it anywhere else beside in richtext that wants to display checkboxes? I think not, so it does not make much sense to not pass it into CheckboxList and benefit from it.

About that _representational widgets_ I worry it is not CWM philosophy ( and I sometimes are not much happy about it ). CWM is more like glue/controller, then just widget. It was always designed as controller, that just specify type of widget ( so also "creates" widget ), but its main goal is init/store methods that sets initial values and store what user specify at the end. And often also do validation.

Maybe it helps if someone outside look at it like @imobachgs or @ancorgs what is their opinion about it and what will fit the best. If we should try dumb widgets or keep consistency with rest of CWM.

Reply:
> I think that details widget can be used standalone without issues.
But can you reuse standlone CheckboxList? I think it does not provide much value. So if I want to reuse it I am force to use MasterDetailSelector, but what if I do not want details? Then I have to write it again from scratch and CheckboxList itself does not provide much.

That's true. Rigth now, `CheckboxList` does not make sense without `MasterDetailSelector`.

> Also consider what provides standalone CheckboxItem. Does it make sense to use it anywhere else beside in richtext that wants to display checkboxes? I think not

Also true. It simple produces a checkbox _simulation_ in a format understandable by a RichText.

> About that representational widgets I worry it is not CWM philosophy ( and I sometimes are not much happy about it ). CWM is more like glue/controller, then just widget. It was always designed as controller, that just specify type of widget ( so also "creates" widget ), but its main goal is init/store methods that sets initial values and store what user specify at the end. And often also do validation.

Thanks for that explanation. I knew is not the CWM philosophy but now I have more details.

> Maybe it helps if someone outside look at it like @imobachgs or @ancorgs what is their opinion about it and what will fit the best. If we should try dumb widgets or keep consistency with rest of CWM.

All opinions are welcome, of course. But I guess it will not pay off only for this screen. Probably we will need more valid scenarios to justify starting to use a new approach. If so, I'll ask for help / pair-programming to fit CWM requirements.

Reply:
I understand both points of view and I would like to start a discussion (not in this PR, of course) about the overall CWM philosophy. Maybe it is time to unify our vision about it.

In `yast2-storage-ng` you can see many widgets having a separate controller, in `yast2-configuration-management` we only use the `#store` method in a single widget, etc. I tend to think that "dump" widgets are easier to reuse... as widgets (not as controllers with all the associated logic). But having a discussion and setting some rules would help to move forward.

What do you think?

Reply:
It is fine for me to unify view about usage especially if we agreed, we can add some helpers to make its usage easier.

Reply:
I also agree that a discussion is needed. IMHO, we need to re-think our UI tools completely, from libyui to CWM. I don't say we have to kill everything, but maybe start looking to other approaches as a possible future improvement.

Regarding this PR, I agree with Josef about not breaking CWM philosophy. If we include something under CWM namespace, I would try to follow the same approach as the rest of existing widgets.

Reply:
So, where/when do you think we should continue talking about that?

Reply:
ideally try to organize meeting with all interested people, create notes from it and ideally result should do design document that will be then implemented. Something like we do team charter :)

Reply:
Well, IIUC in the discussion during the retrospective meeting, the idea is a) to continue with the RichText approach for this new widget, and b) to arrange a specific meeting to discuss about how to improve libyui to avoid this pain.

For a) I guess that POs should decide when to continue with the implementation of this new CWM widget. And for b) we can try proposing a date/hour. But fist, I think we need to write some starting document (gist ?) with the problems discussed and maybe with some possible solution we have already mentioned. What do you think?

Reply:
Yes, you remember correctly. But the problem I see is that this is a different topic. Related but different.

Here we were talking about 

> to re-think our UI tools completely, from libyui to CWM

and more specifically to move forward (or not) the idea of having _dumb_ widgets which only responsibility will be to present the data.

Reply:
After our last retro meeting, I was talking with the PO to see if we could postpone this now that we [fixed the most important problem in `yast2-packager`](https://github.com/yast/yast-packager/pull/511) (the other candidate module to make use of this new set of dialog/widgets). But not, we need to move it forward since we want

* to unify those two dialogs
* to fix the other issues in those selectors

I still believe that a more _state less_ approach would be better to deal with YaST widgets, keeping the business logic outside the widgets, which only should take care of the visual data representation and user interaction.

However, I'd like to finish this endless task as soon as possible. So, I am going to ask you to help me move this forward. Ideally, a pair-programming session to adapt this PR (if possible) or create a new one is quite welcome.

We can continue the discussion later, as Iván pointed out in previous comments.

Thank you so much for your support, reviews, comments and suggestions folks.